### PR TITLE
feat: add piotroski's f-score

### LINKED ISSUE #1: feature: add F-Score in the stock ranking
* Implementing [F-Score](https://www.anderson.ucla.edu/sites/default/files/documents/areas/prg/asam/2019/F-Score.pdf) rating in project.

# Value Investing: Using the Piotroski F-Score to Identify Winning Stocks

This README provides a simplified explanation of Joseph D. Piotroski's research paper titled **"Value Investing: The Use of Historical Financial Statement Information to Separate Winners from Losers."** The paper introduces the **Piotroski F-Score**, a scoring system that helps investors identify undervalued stocks with strong financial health.

---

## Introduction

Value investing involves buying stocks that appear undervalued based on financial ratios like the **book-to-market (BM) ratio**. However, not all high BM stocks are good investments—some are financially distressed and continue to underperform.

**Objective of the Study:**

- **Enhance value investing** by incorporating fundamental financial analysis.
- Develop a simple scoring system (**Piotroski F-Score**) to separate financially strong firms ("winners") from weak ones ("losers") within the high BM category.

---

## The Piotroski F-Score

The **F-Score** is a **9-point scoring system** based on specific financial criteria derived from a company's financial statements. Each criterion adds 1 point if positive, with a maximum score of 9.

### 1. **Profitability Criteria (4 points)**

- **Positive Return on Assets (ROA):** Is the company profitable?
- **Positive Operating Cash Flow (CFO):** Is the company generating cash from operations?
- **Increase in ROA:** Has profitability improved over the past year?
- **Accruals (CFO > ROA):** Are earnings supported by cash flow, indicating high-quality earnings?

### 2. **Leverage, Liquidity, and Source of Funds (3 points)**

- **Decrease in Leverage:** Has the company's debt relative to assets decreased?
- **Increase in Current Ratio:** Has short-term liquidity improved?
- **No New Equity Issuance:** Has the company avoided diluting shareholders by not issuing new shares?

### 3. **Operating Efficiency Criteria (2 points)**

- **Increase in Gross Margin:** Is the company improving its profit margins?
- **Increase in Asset Turnover:** Is the company generating more sales per asset?

---

## How to Calculate the F-Score

For each criterion:

1. **Evaluate the Condition:** Check if the company meets the criterion based on its financial statements.
2. **Assign Points:** Give 1 point for each positive criterion.
3. **Sum the Points:** Add up all the points to get the F-Score (ranging from 0 to 9).

**Interpretation:**

- **High F-Score (8–9):** Strong financial health; likely to outperform.
- **Low F-Score (0–3):** Weak financial health; may underperform.

---

## Applying the F-Score in Investment Strategy

**Step 1:** Identify high BM (value) stocks.

**Step 2:** Calculate the F-Score for each stock.

**Step 3:** Select stocks with high F-Scores for your portfolio.

**Benefits:**

- **Improved Returns:** Historically, high F-Score stocks have outperformed low F-Score stocks.
- **Risk Management:** Avoids financially distressed companies that might appear undervalued but are poor investments.

---

## Research Findings

- **Sample Size:** Analyzed 14,043 high BM firms from 1976 to 1996.
- **Results:** High F-Score firms significantly outperformed low F-Score firms.
- **Consistency:** The strategy was effective across different sizes and industries.

---

## Practical Example

Imagine you're evaluating two companies, both with high BM ratios:

### Company A:

- **F-Score:** 8
- **Indicators:** Positive profitability, decreasing debt, improving margins.

### Company B:

- **F-Score:** 2
- **Indicators:** Negative earnings, increasing debt, declining efficiency.

**Investment Decision:** Choose Company A for your portfolio due to its high F-Score, indicating strong financial health.

---

## Limitations and Considerations

- **Data Availability:** Requires access to detailed financial statements.
- **Market Changes:** The effectiveness may vary with market conditions.
- **Complementary Analysis:** Consider combining with other analyses (e.g., qualitative factors).

---

## Conclusion

The Piotroski F-Score is a valuable tool for investors seeking to enhance their value investing strategy by focusing on financially strong companies. By systematically applying this scoring system, investors can improve portfolio performance and reduce the risk of investing in financially distressed firms.

---

## References

- **Original Research Paper:** [Value Investing: The Use of Historical Financial Statement Information to Separate Winners from Losers](https://www.anderson.ucla.edu/sites/default/files/documents/areas/prg/asam/2019/F-Score.pdf) by Joseph D. Piotroski.

---

### PR DESCRIPTION
**Pull request by naineel1209**
closes #1

## Description

This PR adds Piotroski's F-Score to the repo.

#### 1. **Profitability Ratios**
- **Return on Assets (ROA):**
  - `profitLoss -> Net Profit +`
  - `balanceSheet -> Total Assets`

- **Cash from Operating Activities:**
  - `cashFlows -> Cash from Operating Activity +`

- **Change in ROA:**
  - Same metrics as ROA.

- **Cash from Operating Activities vs. Net Profit:**
  - `cashFlows -> Cash from Operating Activity +`
  - `profitLoss -> Net Profit +`

#### 2. **Leverage, Liquidity, and Source of Funds**
- **Long-term Debt to Total Assets Ratio:**
  - `balanceSheet -> Borrowings +`
  - `balanceSheet -> Total Assets`

- **Current Ratio:**
  - `balanceSheet -> Other Assets +`
  - `balanceSheet -> Other Liabilities +`

- **Equity Capital:**
  - `balanceSheet -> Equity Capital`

#### 3. **Operating Efficiency**
- **Gross Margin (Operating Profit Margin - OPM):**
  - Standard for non-banks: `profitLoss -> OPM %`
  - For banks: `profitLoss -> Net Profit +`, `profitLoss -> Revenue`

- **Asset Turnover Ratio:**
  - Standard for non-banks: `profitLoss -> Sales +`
  - For banks: `profitLoss -> Revenue`
  - `balanceSheet -> Total Assets`

### Response Sample

```json
    {
        "ISIN": "INE787D01026",
        "Industry/Rating": "Auto Components",
        "Market/Fair Value": "118,540.32",
        "Name of the Instrument": "Balkrishna Industries Limited",
        "Percentage of AUM": "1.51%",
        "Quantity": "4,189,074",
        "fScore": "Not Available",
        "marketCap": "Large Cap",
        "marketCapValue": "1,53,485",
        "stockRate": 10.5,
        "url": "https://www.screener.in/company/HINDALCO/consolidated/"
    },
    {
        "ISIN": "INE584A01023",
        "Industry/Rating": "Minerals & Mining",
        "Market/Fair Value": "91,960.07",
        "Name of the Instrument": "NMDC Limited",
        "Percentage of AUM": "1.17%",
        "Quantity": "41,300,671",
        "fScore": 5,
        "marketCap": "Large Cap",
        "marketCapValue": "65,576",
        "stockRate": 11.69,
        "url": "https://www.screener.in/company/NMDC/consolidated/"
    },
    {
        "ISIN": "INE022Q01020",
        "Industry/Rating": "Capital Markets",
        "Market/Fair Value": "90,017.87",
        "Name of the Instrument": "Indian Energy Exchange Limited",
        "Percentage of AUM": "1.15%",
        "Quantity": "44,206,584",
        "fScore": 3,
        "marketCap": "Small Cap",
        "marketCapValue": "15.2",
        "stockRate": 7.19,
        "url": "https://www.screener.in/company/523222/consolidated/"
    },
    {
        "ISIN": "INE010B01027",
        "Industry/Rating": "Pharmaceuticals & Biotechnology",
        "Market/Fair Value": "86,609.72",
        "Name of the Instrument": "Zydus Lifesciences Limited",
        "Percentage of AUM": "1.10%",
        "Quantity": "7,678,847",
        "fScore": "Not Available",
        "marketCap": "Large Cap",
        "marketCapValue": "1,13,493",
        "stockRate": 8.36,
        "url": "https://www.screener.in/company/ZYDUSLIFE/consolidated/"
    }
```

> [!NOTE]
> Some stocks, such as `HDFC Bank`, contain quarterly data when annual data is expected. In such cases, we will skip these stocks and return `'No Data Available'` in the `fScore` field.
>
> image

### CODE REVIEW

File: clients/mongo/mongo_client.go [hunk 1]
```diff
@@ -15,12 +17,21 @@ var (
 )
 
 func init() {
+	// Load environment variables from .env file
+	err := godotenv.Load()
+	if err != nil {
+		log.Println("Error loading .env file")
+	}
```
Comment (shivamsouravjha) [inline clients/mongo/mongo_client.go:24]: dot env can be remove all along(even from porject as not needed)

File: .gitignore [hunk 2]
```diff
@@ -1 +1,27 @@
...
+# Dependency directories (remove the comment below to include it)
+# vendor/
+
+# Go workspace file
+go.work
+go.work.sum
+
+# env file
+.env
+
```
Comment (shivamsouravjha) [inline .gitignore:26]: any reason to add these?

**Comment by shivamsouravjha** [comment]
Hey @naineel1209 would it possible to share any snippet of final outcome? like the final ratio that's shown?

File: .gitignore (hunk 2 above)
Comment (naineel1209) [reply .gitignore:26]: I had this stock gitignore file for golang, so added it to the project.

File: clients/mongo/mongo_client.go [hunk 3]
```diff
@@ -15,13 +16,16 @@ var (
 )
 
 func init() {
+	log.Println("MONGO_URI:", os.Getenv("MONGO_URI"))
+	log.Println("CLOUDINARY_URL:", os.Getenv("CLOUDINARY_URL"))
+
```
Comment (shivamsouravjha) [inline clients/mongo/mongo_client.go:21]: let's use the zap.l.error like used in other code snippet

**Comment by naineel1209** [comment]
Hi @shivamsouravjha, I've added a response sample to the PR comment. As well as resolved the comments.

**Review by shivamsouravjha — APPROVED** [review]
LGTM!

**Pull request merged**